### DISCUSSION TITLE
Support for complex numbers

### DISCUSSION THREAD

**Discussion by OliverAh** [OpenModelica/OpenModelica#12502] [Q&A] [answered]
### Does OpenModelica support complex numbers?

With complex numbers I refer to numbers associated to the square root of negative one, e.g. $z = 0.2 + i*0.5$ with $i = \sqrt{-1}$ (may also see [https://en.wikipedia.org/wiki/Complex_number)](https://en.wikipedia.org/wiki/Complex_number)

I would like to use these to describe state-variables, as well as modify them, and describe interactions between them. This would require to use basic mathematical operations, e.g. addition, multiplication, division, as well as solvers for linear systems of equations.

Is it possible to do this with OpenModelica? 

I was recently introduced to Simcenter Amesim (Siemens) and learned that it is not possible there. That is why I am looking for alternatives now. 

I have not used OpenModelica before. In case it is not possible with the standard scope, I am certainly willing to implement an additional library (hope the terminology is correct), but I would like to confirm first that this would be the way to go :)

Thanks already for your feedback!

**Comment by casella** [comment, marked as answer]
Of course it does: a `Complex` operator record is defined in a library which is loaded by default, which handles basic operators such as `+-*/`. There is also a `Modelica.ComplexMath` library to compute functions taking a complex input argument.

The principles of operator records are described in [Section 14](https://specification.modelica.org/maint/3.6/overloaded-operators.html) of the Modelica Language Specification.

You can check the [PowerGrids](https://github.com/PowerGrids/PowerGrids) library to see how this can be used. It can be installed in OMEdit with the [Package Manager](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omedit.html#omedit-install-library-label).

**Reply by OliverAh** [reply]
Great news, thank you so much!
